### PR TITLE
 Migrate to Swift 4.2

### PR DESCRIPTION
**Pull request by tiagomartinho**

### CODE REVIEW

File: TinyConsole/TinyConsole.swift [hunk 1]
```diff
@@ -14,8 +14,8 @@ open class TinyConsole {
     var consoleController: TinyConsoleController
     
     public static var textAppearance = [
-        NSFontAttributeName: UIFont(name: "Menlo", size: 12.0)!,
-        NSForegroundColorAttributeName: UIColor.white]
+        NSAttributedString.Key.font: UIFont(name: "Menlo", size: 12.0)!,
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsole.swift]: You can also just write .font and .foregroundColor here.

Comment (tiagomartinho) [reply TinyConsole/TinyConsole.swift]: Thanks @ohitsdaniel! I used the migrator and it didn't catch that

File: TinyConsole-Example/MainViewController.swift [hunk 2]
```diff
@@ -47,25 +47,25 @@ class MainViewController: UITableViewController {
     
     func setupNavigationItems() {
         navigationItem.leftBarButtonItems = [
-            UIBarButtonItem(title: "Add Log", style: UIBarButtonItemStyle.plain, target: self, action: #selector(addLog))
+            UIBarButtonItem(title: "Add Log", style: UIBarButtonItem.Style.plain, target: self, action: #selector(addLog))
```
Comment (ohitsdaniel) [inline TinyConsole-Example/MainViewController.swift]: Just `.plain`.

File: TinyConsole-Example/MainViewController.swift [hunk 3]
```diff
@@ -16,7 +16,7 @@ class MainViewController: UITableViewController {
     init() {
         tableViewDelegate = MainTableViewDelegate()
         tableViewDataSource = MainTableViewDataSource()
-        super.init(style: UITableViewStyle.plain)
+        super.init(style: UITableView.Style.plain)
```
Comment (ohitsdaniel) [inline TinyConsole-Example/MainViewController.swift]: `.plain`

File: TinyConsole-Example/AppDelegate.swift [hunk 4]
```diff
@@ -13,12 +13,12 @@ import TinyConsole
...
     var window: UIWindow?
     
-    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         window = UIWindow(frame: UIScreen.main.bounds)
         
         let viewController = UINavigationController(rootViewController: MainViewController())
         viewController.title = "Main"
-        viewController.tabBarItem = UITabBarItem(tabBarSystemItem: UITabBarSystemItem.recents, tag: 0)
+        viewController.tabBarItem = UITabBarItem(tabBarSystemItem: UITabBarItem.SystemItem.recents, tag: 0)
```
Comment (ohitsdaniel) [inline TinyConsole-Example/AppDelegate.swift]: `.recents`

File: TinyConsole-Example/MainViewController.swift [hunk 5]
```diff
@@ -47,25 +47,25 @@ class MainViewController: UITableViewController {
     
     func setupNavigationItems() {
         navigationItem.leftBarButtonItems = [
-            UIBarButtonItem(title: "Add Log", style: UIBarButtonItemStyle.plain, target: self, action: #selector(addLog))
+            UIBarButtonItem(title: "Add Log", style: UIBarButtonItem.Style.plain, target: self, action: #selector(addLog))
         ]
         navigationItem.rightBarButtonItems = [
-            UIBarButtonItem( title: "Add Marker", style: UIBarButtonItemStyle.plain, target: self, action: #selector(addMarker)),
-            UIBarButtonItem( title: "Clear", style: UIBarButtonItemStyle.plain, target: self, action: #selector(clear)),
+            UIBarButtonItem( title: "Add Marker", style: UIBarButtonItem.Style.plain, target: self, action: #selector(addMarker)),
```
Comment (ohitsdaniel) [inline TinyConsole-Example/MainViewController.swift]: Same here.

File: TinyConsole/TinyConsole.swift [hunk 6]
```diff
@@ -60,7 +60,7 @@ open class TinyConsole {
         
         // set standard text appearance and override foreground color attribute
         formattedText.addAttributes(TinyConsole.textAppearance, range: range)
-        formattedText.addAttribute(NSForegroundColorAttributeName, value: color, range: range)
+        formattedText.addAttribute(NSAttributedString.Key.foregroundColor, value: color, range: range)
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsole.swift]: `.foregroundColor` should be fine here

File: TinyConsole/TinyConsoleController.swift [hunk 7]
```diff
@@ -89,8 +89,8 @@ open class TinyConsoleController: UIViewController {
         setupConstraints()
     }
     
-    open override func motionBegan(_ motion: UIEventSubtype, with event: UIEvent?) {
-        if (motion == UIEventSubtype.motionShake) {
+    open override func motionBegan(_ motion: UIEvent.EventSubtype, with event: UIEvent?) {
+        if (motion == UIEvent.EventSubtype.motionShake) {
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleController.swift]: `.motionShake`

File: TinyConsole/TinyConsoleController.swift [hunk 8]
```diff
@@ -103,19 +103,19 @@ open class TinyConsoleController: UIViewController {
         for view in view.subviews {
             view.removeFromSuperview()
         }
-        for childViewController in childViewControllers {
-            childViewController.removeFromParentViewController()
+        for childViewController in children {
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleController.swift]: Just a preference from my side, but could you change this to

```
children.forEach {
    $0.removeFromParent()
}
```

File: TinyConsole/TinyConsoleViewController.swift [hunk 9]
```diff
@@ -40,20 +40,20 @@ class TinyConsoleViewController: UIViewController {
         }
     }
     
-    func customText(sender: UITapGestureRecognizer) {
-        let alert = UIAlertController(title: "Custom Log", message: "Enter text you want to log.", preferredStyle: UIAlertControllerStyle.alert)
+    @objc func customText(sender: UITapGestureRecognizer) {
+        let alert = UIAlertController(title: "Custom Log", message: "Enter text you want to log.", preferredStyle: UIAlertController.Style.alert)
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.alert`

File: TinyConsole/TinyConsoleViewController.swift [hunk 10]
```diff
@@ -40,20 +40,20 @@ class TinyConsoleViewController: UIViewController {
...
         alert.addTextField { (textField: UITextField) in
             textField.keyboardType = .alphabet
         }
         
-        let okAction = UIAlertAction(title: "Add log", style: UIAlertActionStyle.default) {
+        let okAction = UIAlertAction(title: "Add log", style: UIAlertAction.Style.default) {
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.default`

File: TinyConsole/TinyConsoleViewController.swift [hunk 11]
```diff
@@ -40,20 +40,20 @@ class TinyConsoleViewController: UIViewController {
...
             (action: UIAlertAction) in
             if let text = alert.textFields?.first?.text, !text.isEmpty {
                 TinyConsole.print(text)
             }
         }
         
-        let cancelAction = UIAlertAction(title: "Cancel", style: UIAlertActionStyle.cancel, handler: nil)
+        let cancelAction = UIAlertAction(title: "Cancel", style: UIAlertAction.Style.cancel, handler: nil)
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.cancel`

File: TinyConsole/TinyConsoleViewController.swift [hunk 12]
```diff
@@ -84,12 +84,12 @@ class TinyConsoleViewController: UIViewController {
         view.addGestureRecognizer(showAdditionalActionsGesture)
     }
     
-    func additionalActions(sender: UITapGestureRecognizer) {
+    @objc func additionalActions(sender: UITapGestureRecognizer) {
         
-        let alert = UIAlertController(title: nil, message: nil, preferredStyle: UIAlertControllerStyle.actionSheet)
+        let alert = UIAlertController(title: nil, message: nil, preferredStyle: UIAlertController.Style.actionSheet)
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.actionSheet`

File: TinyConsole/TinyConsoleViewController.swift [hunk 13]
```diff
@@ -84,12 +84,12 @@ class TinyConsoleViewController: UIViewController {
...
         
         let defaultActions = [
-            UIAlertAction(title: "Send Email", style: UIAlertActionStyle.default) {
+            UIAlertAction(title: "Send Email", style: UIAlertAction.Style.default) {
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.default`

File: TinyConsole/TinyConsoleViewController.swift [hunk 14]
```diff
@@ -102,7 +102,7 @@ class TinyConsoleViewController: UIViewController {
                     self.present(composeViewController, animated: true, completion: nil)
                 }
             },
-            UIAlertAction(title: "Clear", style: UIAlertActionStyle.destructive) {
+            UIAlertAction(title: "Clear", style: UIAlertAction.Style.destructive) {
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.destructive`

File: TinyConsole/TinyConsoleViewController.swift [hunk 15]
```diff
@@ -112,13 +112,13 @@ class TinyConsoleViewController: UIViewController {
         }
         
         
-        let cancelAction = UIAlertAction(title: "Cancel", style: UIAlertActionStyle.cancel, handler: nil)
+        let cancelAction = UIAlertAction(title: "Cancel", style: UIAlertAction.Style.cancel, handler: nil)
```
Comment (ohitsdaniel) [inline TinyConsole/TinyConsoleViewController.swift]: `.cancel`

**Comment by tiagomartinho** [comment]
Thanks @ohitsdaniel !

Let me know if I missed something.

I also formatted the code to be more consistent across the project and added what I think was a misuse of the UIKit API for adding/removing container view controllers.

Let me know what you think

**Comment by warpling** [comment]
Bump!

**Comment by Cosmo** [comment]
Thanks @tiagomartinho, @freegor, @ohitsdaniel for the pull request, changes, push and patience!

**Pull request merged**